### PR TITLE
 Remove Python virtual environment from Dockerfile

### PR DESCRIPTION
**Pull request by snaselj**
# Closes: #4333

Tested `final-dev` image with `nautobot-plugin-circuits-maintenance` `u/snaselj-next-2.0` branch. All tests passed there.

# What's Changed

- Removed Python virtual environment from Docker image.
- Replaced `POETRY_PARALLEL` with `POETRY_INSTALLER_PARALLEL`.
- Added `poetry` stub stage for poetry installation and configuration.
- Added `invoke cli` `--root` argument to run CLI as root.
- Added `invoke build` `--service` argument to be able to specify a service to build.

### CODE REVIEW

File: docker/Dockerfile [hunk 1]
```diff
@@ -71,11 +71,17 @@ RUN --mount=type=cache,target="/root/.cache/pip",sharing=locked \
...
 RUN mkdir /opt/nautobot /opt/nautobot/.cache /prom_cache /source && \
     groupadd --gid 999 --system nautobot && \
     useradd --uid 999 --gid 999 --system --shell /bin/bash --create-home --home-dir /opt/nautobot nautobot && \
-    chown -R nautobot:nautobot /opt/nautobot /prom_cache /source
+    chown -R nautobot:nautobot \
+    /opt/nautobot \
+    /prom_cache \
+    /source \
+    /usr/local/bin \
+    /usr/local/lib/python${PYTHON_VER}/site-packages
```
Comment (nniehoff) [inline docker/Dockerfile:84]: This seems like a security risk allowing the execution user to write to `site-packages` and bin, the point of not running as root is to not allow this specifically

**Review by nniehoff — DISMISSED** [review]
This seems to work for me, I've got some issues later down the pipeline I'll chase down later but the container seems to work.  I am concerned however with the `chown nautobot /usr/local/bin .../site-packages` I believe this introduces a security risk

File: docker/Dockerfile (hunk 1 above)
Comment (bryanculver) [reply docker/Dockerfile:84]: @snaselj @nniehoff Can/should we run the `pip` or `poetry` commands as `root` then?

Comment (snaselj) [reply docker/Dockerfile:84]: There are another places where user `nautobot` has write access in `final` image, should I fix it as well?

```
COPY --from=build-nautobot --chown=nautobot:nautobot /source/dist /source/dist
COPY --from=final-dev --chown=nautobot:nautobot /opt/nautobot/nautobot_config.py /opt/nautobot/nautobot_config.py
COPY --from=final-dev --chown=nautobot:nautobot /opt/nautobot/ui /opt/nautobot/ui
COPY --chown=nautobot:nautobot docker/uwsgi.ini /opt/nautobot
```

Comment (bryanculver) [reply docker/Dockerfile:84]: I think this is a slippery slope: by @nniehoff’s comments none of this should be writable but we need to create the docs and UI so maybe that needs to be run as root too. However everything starts with the site-packages so let’s start with the Python executables not being writable.

If this is permitted in the Venv install method, why not here?

Comment (nniehoff) [reply docker/Dockerfile:84]: > @snaselj @nniehoff Can/should we run the `pip` or `poetry` commands as `root` then?

During the build phase yes they should I would just use the `USER` directive to switch back and forth

File: docker/Dockerfile [hunk 2]
```diff
@@ -259,13 +260,10 @@ FROM system-dependencies AS final
...
-ENV VIRTUAL_ENV="/opt/nautobot"
 
-COPY --from=python-dependencies --chown=nautobot:nautobot /opt/nautobot/bin /opt/nautobot/bin
-COPY --from=python-dependencies --chown=nautobot:nautobot /opt/nautobot/include /opt/nautobot/include
-COPY --from=python-dependencies --chown=nautobot:nautobot /opt/nautobot/lib /opt/nautobot/lib
-COPY --from=python-dependencies --chown=nautobot:nautobot /opt/nautobot/lib64 /opt/nautobot/lib64
-COPY --from=python-dependencies --chown=nautobot:nautobot /opt/nautobot/pyvenv.cfg /opt/nautobot/pyvenv.cfg
+ARG PYTHON_VER
+COPY --from=python-dependencies /usr/local/lib/python${PYTHON_VER}/site-packages /usr/local/lib/python${PYTHON_VER}/site-packages
+COPY --from=python-dependencies /usr/local/bin /usr/local/bin
```
[the remark this answers is not in the document]
Comment (bryanculver) [reply docker/Dockerfile:266]: @snaselj

[the remark this answers is not in the document]
Comment (snaselj) [reply docker/Dockerfile:258]: Verified on `final` image:

```
nautobot@b369e73e781f:/usr/local/bin$ ls -Al |grep nautobot
-rwxr-xr-x 1 root root   221 Sep  4 13:27 nautobot-server
nautobot@b369e73e781f:/usr/local/bin$ 
nautobot@b369e73e781f:/usr/local/bin$ cd ../lib/python3.8/
nautobot@b369e73e781f:/usr/local/lib/python3.8$ ls -Al |grep nautobot
nautobot@b369e73e781f:/usr/local/lib/python3.8$ cd site-packages/
nautobot@b369e73e781f:/usr/local/lib/python3.8/site-packages$ ls -Al |grep nautobot
drwxr-xr-x 1 root root     240 Sep  4 13:27 nautobot
drwxr-xr-x 1 root root     136 Sep  4 13:27 nautobot-2.0.0rc2.dist-info
nautobot@b369e73e781f:/usr/local/lib/python3.8/site-packages$ 
```

File: docker/Dockerfile (hunk 1 above)
Comment (snaselj) [reply docker/Dockerfile]: All pip and poetry commands are run as root now.

**Comment by snaselj** [comment]
Rebased on top of the latest `next`.

**Comment by bryanculver** [comment]
@nniehoff Can you check @snaselj's updates?

**Review by gsnider2195 — APPROVED** [review]

**Pull request merged**